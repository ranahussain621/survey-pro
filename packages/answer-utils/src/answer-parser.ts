import { CURRENCY_SYMBOLS } from './consts'
import { Answer, FullNameValue, ServerSidePaymentValue } from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'
import Big from 'big.js'

function fileUpload(answer: Answer, livePreview = false) {
  if (livePreview) {
    return {
      filename: (answer as any).name
    }
  }

  if (helper.isURL(answer.value)) {
    return {
      filename: '',
      url: answer.value
    }
  }

  return {
    filename: answer.value.filename || '',
    url: answer.value.url
  }
}

function rating(answer: Answer): string {
  return answer.value
}

function singleChoice(answer: Answer): string {
  const choice = answer.properties!.choices?.find(row => row.id === answer.value)
  return choice ? choice.label : ''
}

function multipleChoice(answer: Answer): string {
  return (answer.properties?.choices?.filter(row => answer.value.value.includes(row.id)) || [])
    .map(row => row.label)
    .concat([answer.value.other])
    .filter(row => helper.isValid(row))
    .join(', ')
}

function fullName(answer: Answer): FullNameValue {
  return answer.value
}

function address(answer: Answer): string {
  return [
    answer.value.address1,
    ',',
    answer.value.address2,
    answer.value.city,
    ',',
    answer.value.state,
    ',',
    answer.value.country,
    answer.value.zip
  ]
    .filter(Boolean)
    .join(' ')
}

function legalTerms(answer: Answer): string {
  return helper.isTrue(answer.value) ? 'Yes' : 'No'
}

function dateRange(answer: Answer): string {
  return [answer.value.start, answer.value.end].filter(Boolean).join(' - ')
}

function inputTable(answer: Answer): string {
  const columns = answer.properties?.tableColumns

  if (helper.isValidArray(columns)) {
    const result: string[] = []

    answer.value.forEach((values: Record<string, string>) => {
      if (helper.isValid(values)) {
        const row = columns!.map(column => values[column.id]).join(', ')
        result.push(row)
      }
    })

    return result.join('\n')
  }

  return ''
}

function payment(answer: Answer): string {
  const value = answer.value as ServerSidePaymentValue
  const price = Big(value.amount).div(100).toFixed(2)
  let result = CURRENCY_SYMBOLS[value.currency] + price

  if (helper.isValid(value.paymentIntentId)) {
    result = `Succeeded ${result}`
  } else {
    result = `Incomplete ${result}`
  }

  return result
}

export default {
  fileUpload,
  rating,
  singleChoice,
  multipleChoice,
  fullName,
  address,
  legalTerms,
  dateRange,
  inputTable,
  payment
}
