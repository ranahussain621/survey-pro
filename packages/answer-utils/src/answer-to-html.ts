import parser from './answer-parser'
import { Answer, FieldKindEnum, QUESTION_FIELD_KINDS } from '@heyform-inc/shared-types-enums'

export function answersToHtml(answers: Answer[]): string {
  const html = answers
    .map(answer => {
      const value = parseHtmlAnswer(answer)

      return `
<li>
  <h3>${answer.title}</h3>
  <p>${value}</p>
</li>
`
    })
    .join('')

  return `<ol>${html}</ol>`
}

function parseHtmlAnswer(answer: Answer): string {
  let value = ''

  if (QUESTION_FIELD_KINDS.includes(answer.kind)) {
    switch (answer.kind) {
      case FieldKindEnum.FILE_UPLOAD:
        const file = parser.fileUpload(answer)
        value = `${file.filename} (${file.url})`
        break

      case FieldKindEnum.RATING:
      case FieldKindEnum.OPINION_SCALE:
        value = parser.rating(answer)
        break

      case FieldKindEnum.YES_NO:
        value = parser.singleChoice(answer)
        break

      case FieldKindEnum.MULTIPLE_CHOICE:
      case FieldKindEnum.PICTURE_CHOICE:
        value = parser.multipleChoice(answer)
        break

      case FieldKindEnum.FULL_NAME:
        const name = parser.fullName(answer)
        value = `${name.firstName} ${name.lastName}`
        break

      case FieldKindEnum.ADDRESS:
        value = parser.address(answer)
        break

      case FieldKindEnum.LEGAL_TERMS:
        value = parser.legalTerms(answer)
        break

      case FieldKindEnum.DATE_RANGE:
        value = parser.dateRange(answer)
        break

      case FieldKindEnum.INPUT_TABLE:
        value = parser.inputTable(answer)
        break

      case FieldKindEnum.PAYMENT:
        value = parser.payment(answer)
        break

      default:
        value = answer.value?.toString()
    }
  }

  return value
}
