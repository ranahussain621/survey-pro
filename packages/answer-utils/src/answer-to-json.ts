import parser from './answer-parser'
import { Answer, FieldKindEnum } from '@heyform-inc/shared-types-enums'

interface AnswersToJsonOptions {
  plain?: boolean
}

export function answersToJson(
  answers: Answer[],
  options?: AnswersToJsonOptions
): Record<string, any> {
  const result: Record<string, any> = {}

  answers.forEach(answer => {
    result[answer.id] = parseJsonAnswer(answer, options?.plain)
  })

  return result
}

function parseJsonAnswer(answer: Answer, plain = false): any {
  let value: any

  switch (answer.kind) {
    case FieldKindEnum.FILE_UPLOAD:
      value = parser.fileUpload(answer)
      if (plain) {
        value = `${value.filename} (${value.url})`
      }
      break

    case FieldKindEnum.RATING:
    case FieldKindEnum.OPINION_SCALE:
      value = parser.rating(answer)
      break

    case FieldKindEnum.YES_NO:
      value = parser.singleChoice(answer)
      break

    case FieldKindEnum.MULTIPLE_CHOICE:
    case FieldKindEnum.PICTURE_CHOICE:
      value = parser.multipleChoice(answer)
      break

    case FieldKindEnum.FULL_NAME:
      value = parser.fullName(answer)
      if (plain) {
        value = `${value.firstName} ${value.lastName}`
      }
      break

    case FieldKindEnum.ADDRESS:
      value = answer.value
      if (plain) {
        value = parser.address(answer)
      }
      break

    case FieldKindEnum.LEGAL_TERMS:
      value = parser.legalTerms(answer)
      break

    case FieldKindEnum.DATE_RANGE:
      value = answer.value
      if (plain) {
        value = parser.dateRange(answer)
      }
      break

    case FieldKindEnum.INPUT_TABLE:
      value = answer.value
      if (plain) {
        value = parser.inputTable(answer)
      }
      break

    case FieldKindEnum.PAYMENT:
      value = answer.value
      if (plain) {
        value = parser.payment(answer)
      }
      break

    default:
      value = answer.value?.toString()
  }

  return value
}
