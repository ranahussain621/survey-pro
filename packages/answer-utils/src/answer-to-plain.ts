import parser from './answer-parser'
import { Answer, FieldKindEnum } from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

export function answersToPlain(answers: Answer[]): string {
  return answers
    .map(answer => {
      const value = parsePlainAnswer(answer)
      return `${answer.title}\n${value}`
    })
    .join('\n\n')
}

export function parsePlainAnswer(answer: Answer, livePreview = false): string {
  let value: string

  switch (answer.kind) {
    case FieldKindEnum.FILE_UPLOAD:
      const file = parser.fileUpload(answer, livePreview)
      value = livePreview ? file.filename : `${file.filename} (${file.url})`
      break

    case FieldKindEnum.RATING:
    case FieldKindEnum.OPINION_SCALE:
      value = parser.rating(answer)
      break

    case FieldKindEnum.YES_NO:
      value = parser.singleChoice(answer)
      break

    case FieldKindEnum.MULTIPLE_CHOICE:
    case FieldKindEnum.PICTURE_CHOICE:
      value = parser.multipleChoice(answer)
      break

    case FieldKindEnum.FULL_NAME:
      const name = parser.fullName(answer)
      value = `${name.firstName} ${name.lastName}`
      break

    case FieldKindEnum.ADDRESS:
      value = parser.address(answer)
      break

    case FieldKindEnum.LEGAL_TERMS:
      value = parser.legalTerms(answer)
      break

    case FieldKindEnum.DATE_RANGE:
      value = parser.dateRange(answer)
      break

    case FieldKindEnum.INPUT_TABLE:
      value = parser.inputTable(answer)
      break

    case FieldKindEnum.PAYMENT:
      value = parser.payment(answer)
      break

    default:
      value = answer.value?.toString()
  }

  if (livePreview && helper.isEmpty(value)) {
    value = '_____'
  }

  return value
}
