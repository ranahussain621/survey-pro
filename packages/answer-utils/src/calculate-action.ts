import { isNumber } from './helper'
import {
  AnswerValue,
  CalculateEnum,
  NumberCalculateAction,
  StringCalculateAction,
  Variable
} from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

export function calculateAction(
  action: NumberCalculateAction | StringCalculateAction,
  parameters?: Variable[],
  data?: Record<string, string | number>,
  values?: Record<string, AnswerValue>
) {
  if (helper.isEmpty(parameters) || helper.isEmpty(data) || helper.isNil(data![action.variable])) {
    return data || {}
  }

  const variable = parameters!.find(v => v.id === action.variable)!
  let current = data![action.variable]

  switch (variable.kind) {
    case 'string':
      current = calculateString(current as string, action as StringCalculateAction, values)
      break

    case 'number':
      current = calculateNumber(current as number, action as NumberCalculateAction, values)
      break
  }

  return {
    ...data,
    [action.variable]: current
  }
}

function calculateString(
  value: string,
  action: StringCalculateAction,
  values?: Record<string, AnswerValue>
): string {
  let newValue = action.value as string

  if (action.ref) {
    newValue = values?.[action.ref] as string
  }

  if (!helper.isNil(newValue)) {
    switch (action.operator) {
      case CalculateEnum.ADDITION:
        return value + newValue!

      case CalculateEnum.ASSIGNMENT:
        return newValue!
    }
  }

  return value
}

function calculateNumber(
  value: number,
  action: NumberCalculateAction,
  values?: Record<string, AnswerValue>
): number {
  let newValue = action.value as number

  if (action.ref) {
    newValue = values?.[action.ref] as number
  }

  if (!helper.isNil(newValue) && isNumber(newValue!)) {
    switch (action.operator) {
      case CalculateEnum.ADDITION:
        return value + newValue!

      case CalculateEnum.SUBTRACTION:
        return value - newValue!

      case CalculateEnum.MULTIPLICATION:
        return value * newValue!

      case CalculateEnum.DIVISION:
        return newValue! !== 0 ? value / newValue! : value

      case CalculateEnum.ASSIGNMENT:
        return newValue!
    }
  }

  return value
}
