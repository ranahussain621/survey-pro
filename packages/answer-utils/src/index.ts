export * from './answer-to-html'
export * from './answer-to-api-object'
export * from './answer-to-plain'
export * from './answer-to-json'
export * from './field-values-to-answer'
export * from './fields-to-validate-rules'
export * from './validate'
export * from './html-utils'
export * from './flatten-fields'
export * from './validate-condition'
export * from './validate-payload'
export * from './apply-logic-to-fields'
export * from './consts'
export * from './helper'
