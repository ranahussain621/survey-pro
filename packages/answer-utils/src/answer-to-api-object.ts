import { parsePlainAnswer } from './answer-to-plain'
import { Answer } from '@heyform-inc/shared-types-enums'

export function answersToApiObject(answers: Answer[]): Record<string, any> {
  const result: Record<string, any> = {}

  answers.forEach(answer => {
    const key = `(ID: ${answer.id}) ${answer.title}`
    result[key] = parsePlainAnswer(answer)
  })

  return result
}
