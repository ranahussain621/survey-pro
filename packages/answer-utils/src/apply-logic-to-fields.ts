import { calculateAction } from './calculate-action'
import { validateRequiredField } from './validate'
import { validateCondition } from './validate-condition'
import {
  ActionEnum,
  AnswerValue,
  FieldKindEnum,
  FormField,
  Logic,
  NavigateAction,
  NumberCalculateAction,
  QUESTION_FIELD_KINDS,
  StringCalculateAction,
  Variable
} from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

interface IFormField extends FormField {
  parent?: FormField
  isTouched?: boolean
}

export interface PickAndCalcFieldsResult {
  fields: IFormField[]
  variables: Record<string, AnswerValue>
}

function indexFields(fields: IFormField[]) {
  const parents = fields.filter(f => !f.parent && QUESTION_FIELD_KINDS.includes(f.kind))
  const children = fields.filter(f => f.parent && QUESTION_FIELD_KINDS.includes(f.kind))

  let index = 1
  const childrenIndexes: Record<string, number> = {}

  parents.forEach(f => {
    f.index = index++
  })

  children.forEach(f => {
    const parentId = f.parent!.id
    const parent = parents.find(p => p.id === parentId)

    if (!childrenIndexes[parentId]) {
      childrenIndexes[parentId] = 1
    }

    f.parent = parent
    f.index = childrenIndexes[parentId]++
  })
}

export function applyLogicToFields(
  fields?: IFormField[],
  logics?: Logic[],
  parameters?: Variable[],
  values?: Record<string, AnswerValue>
): PickAndCalcFieldsResult {
  const result: PickAndCalcFieldsResult = {
    fields: [],
    variables: {}
  }

  if (helper.isEmpty(fields)) {
    return result
  }

  if (helper.isValid(parameters)) {
    for (const variable of parameters!) {
      result.variables[variable.id] = variable.value
    }
  }

  if (helper.isEmpty(logics) || helper.isEmpty(values)) {
    // Add indexes to every field
    indexFields(fields!)

    return {
      ...result,
      fields: fields!.filter(f => f.kind !== FieldKindEnum.THANK_YOU)
    }
  }

  let index = 0

  while (index < fields!.length) {
    const field = fields![index]

    if (field.kind === FieldKindEnum.THANK_YOU) {
      break
    }

    let isNavigateValidated = false
    const logic = logics!.find(l => l.fieldId === field.id)

    // Add parent field if needed to present SubGroup fields
    if (field.parent && result.fields.findIndex(f => f.id === field.parent!.id) < 0) {
      result.fields.push(field.parent)
    }

    if (logic) {
      const { payloads } = logic
      const calculates = payloads.filter(p => p.action.kind === ActionEnum.CALCULATE)

      for (const calculate of calculates) {
        const { action, condition } = calculate
        const isValidated = validateCondition(field, condition, values)

        if (isValidated) {
          result.variables = calculateAction(
            action as NumberCalculateAction | StringCalculateAction,
            parameters,
            result.variables,
            values
          )
        }
      }

      const navigates = payloads.filter(p => p.action.kind === ActionEnum.NAVIGATE)

      // Find the first navigate which meets the jump rule
      for (const navigate of navigates) {
        const { action, condition } = navigate

        // Whether the jump logic in this field has been verified or not
        field.isTouched = validateCondition(field, condition, values)

        if (field.isTouched) {
          const jumpFieldId = (action as NavigateAction).fieldId
          const jumpIndex = fields!.findIndex(f => f.id === jumpFieldId)
          const isExists = !!result.fields.find(f => f.id === jumpFieldId)

          // Logic jump only works when sending the respondent forward in the form
          if (!isExists && jumpIndex > index) {
            index = jumpIndex
            isNavigateValidated = true

            result.fields.push(field)
            break
          }
        } else {
          field.isTouched = validateRequiredField(field, values)
        }
      }
    }

    if (!isNavigateValidated) {
      index += 1
      result.fields.push(field)
    }
  }

  // Add indexes to every field
  indexFields(result.fields)

  return result
}
