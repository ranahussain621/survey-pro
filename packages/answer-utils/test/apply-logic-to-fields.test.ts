import { applyLogicToFields } from '../src'
import {
  ActionEnum,
  CalculateEnum,
  ComparisonEnum,
  FieldKindEnum,
  FormField,
  Logic,
  Variable
} from '@heyform-inc/shared-types-enums'

const fields: FormField[] = [
  {
    id: 'id_short_text',
    title: 'Short Text',
    kind: FieldKindEnum.SHORT_TEXT
  },
  {
    id: 'id_statement',
    title: 'Statement',
    kind: FieldKindEnum.STATEMENT
  },
  {
    id: 'id_number',
    title: 'Number',
    kind: FieldKindEnum.NUMBER
  },
  {
    id: 'id_statement2',
    title: 'Statement2',
    kind: FieldKindEnum.STATEMENT
  },
  {
    id: 'id_date',
    title: 'Date',
    kind: FieldKindEnum.DATE
  }
]

const logics: Logic[] = [
  {
    fieldId: 'id_short_text',
    payloads: [
      {
        id: 'p1',
        condition: {
          comparison: ComparisonEnum.CONTAINS,
          expected: 'kitty'
        },
        action: {
          kind: ActionEnum.NAVIGATE,
          fieldId: fields[2].id
        }
      }
    ]
  },
  {
    fieldId: 'id_number',
    payloads: [
      {
        id: 'p2',
        condition: {
          comparison: ComparisonEnum.EQUAL,
          expected: 5
        },
        action: {
          kind: ActionEnum.NAVIGATE,
          fieldId: fields[fields.length - 1].id
        }
      }
    ]
  }
]

const variables: Variable[] = [
  {
    id: 'number_var',
    name: 'Score',
    kind: 'number',
    value: 0
  },
  {
    id: 'string_var',
    name: 'Name',
    kind: 'string',
    value: 'Your name is: '
  }
]

test('apply to fields with matched values', () => {
  const result = applyLogicToFields(fields, logics, undefined, {
    id_short_text: 'hello kitty',
    id_number: 5
  })

  expect(result.fields.map(f => f.id)).toStrictEqual(
    fields.filter(f => f.kind !== FieldKindEnum.STATEMENT).map(f => f.id)
  )
})

test('apply to fields with matched text value', () => {
  const result = applyLogicToFields(fields, logics, undefined, {
    id_short_text: 'hello kitty'
  })

  expect(result.fields.map(f => f.id)).toStrictEqual(
    fields.filter(f => f.id !== 'id_statement').map(f => f.id)
  )
})

test('apply to fields with matched number value', () => {
  const result = applyLogicToFields(fields, logics, undefined, {
    id_number: 5
  })

  expect(result.fields.map(f => f.id)).toStrictEqual(
    fields.filter(f => f.id !== 'id_statement2').map(f => f.id)
  )
})

test('apply to fields with unmatched values', () => {
  const result = applyLogicToFields(fields, logics, undefined, {
    id_short_text: 'hello world',
    id_number: 10
  })
  expect(result.fields.map(f => f.id)).toStrictEqual(fields.map(f => f.id))
})

test('apply to fields with empty values', () => {
  const result = applyLogicToFields(fields, logics)
  expect(result.fields.map(f => f.id)).toStrictEqual(fields.map(f => f.id))
})

test('apply to fields without logics', () => {
  const result = applyLogicToFields(fields, undefined, undefined, {
    id_short_text: 'hello kitty'
  })
  expect(result.fields.map(f => f.id)).toStrictEqual(fields.map(f => f.id))
})

test('apply to fields with empty fields', () => {
  const result = applyLogicToFields([], logics, undefined, {
    id_short_text: 'hello world',
    id_number: 10
  })
  expect(result.fields.map(f => f.id)).toStrictEqual([])
})

test('apply to fields with circular navigate logics', () => {
  const _logics: Logic[] = [
    logics[0],
    {
      fieldId: 'id_number',
      payloads: [
        {
          id: 'p2',
          condition: {
            comparison: ComparisonEnum.EQUAL,
            expected: 5
          },
          action: {
            kind: ActionEnum.NAVIGATE,
            fieldId: fields[0].id
          }
        }
      ]
    }
  ]

  const result = applyLogicToFields(fields, _logics, undefined, {
    id_short_text: 'hello kitty',
    id_number: 5
  })
  expect(result.fields.map(f => f.id)).toStrictEqual(
    fields.filter(f => f.id !== 'id_statement').map(f => f.id)
  )
})

test('apply to fields with variables', () => {
  const _logics: Logic[] = [
    {
      fieldId: 'id_short_text',
      payloads: [
        {
          id: 'p1',
          condition: {
            comparison: ComparisonEnum.CONTAINS,
            expected: 'kitty'
          },
          action: {
            kind: ActionEnum.NAVIGATE,
            fieldId: fields[2].id
          }
        },
        {
          id: 'p2',
          condition: {
            comparison: ComparisonEnum.ENDS_WITH,
            expected: 'kitty'
          },
          action: {
            kind: ActionEnum.CALCULATE,
            variable: 'number_var',
            operator: CalculateEnum.ADDITION,
            value: 2
          }
        },
        {
          id: 'p3',
          condition: {
            comparison: ComparisonEnum.STARTS_WITH,
            expected: 'hello'
          },
          action: {
            kind: ActionEnum.CALCULATE,
            variable: 'string_var',
            operator: CalculateEnum.ADDITION,
            ref: 'id_short_text'
          }
        },
        {
          id: 'p5',
          condition: {
            comparison: ComparisonEnum.IS_NOT_EMPTY
          },
          action: {
            kind: ActionEnum.CALCULATE,
            variable: 'string_var2',
            operator: CalculateEnum.ASSIGNMENT,
            value: 'catting'
          }
        }
      ]
    },
    {
      fieldId: 'id_number',
      payloads: [
        {
          id: 'p4',
          condition: {
            comparison: ComparisonEnum.EQUAL,
            expected: 5
          },
          action: {
            kind: ActionEnum.CALCULATE,
            variable: 'number_var',
            operator: CalculateEnum.MULTIPLICATION,
            value: 5
          }
        }
      ]
    }
  ]
  const result = applyLogicToFields(fields, _logics, variables, {
    id_short_text: 'hello kitty',
    id_number: 5
  })
  expect(result.fields.map(f => f.id)).toStrictEqual(
    fields.filter(f => f.id !== 'id_statement').map(f => f.id)
  )
  expect(result.variables).toStrictEqual({
    number_var: 10,
    string_var: 'Your name is: hello kitty'
  })
})

test('apply to group fields', () => {
  const _fields: any[] = [
    {
      id: 'id_short_text',
      title: 'Short Text',
      kind: FieldKindEnum.SHORT_TEXT
    },
    // Group start
    {
      id: 'id_group',
      title: 'Group',
      kind: FieldKindEnum.GROUP
    },
    {
      id: 'id_statement',
      title: 'Statement',
      kind: FieldKindEnum.STATEMENT,
      parent: {
        id: 'id_group',
        title: 'Group',
        kind: FieldKindEnum.GROUP
      }
    },
    {
      id: 'id_number',
      title: 'Number',
      kind: FieldKindEnum.NUMBER,
      parent: {
        id: 'id_group',
        title: 'Group',
        kind: FieldKindEnum.GROUP
      }
    },
    {
      id: 'id_statement2',
      title: 'Statement2',
      kind: FieldKindEnum.STATEMENT,
      parent: {
        id: 'id_group',
        title: 'Group',
        kind: FieldKindEnum.GROUP
      }
    },
    // Group end
    {
      id: 'id_date',
      title: 'Date',
      kind: FieldKindEnum.DATE
    }
  ]
  const _logics: Logic[] = [
    {
      fieldId: 'id_short_text',
      payloads: [
        {
          id: 'p1',
          condition: {
            comparison: ComparisonEnum.CONTAINS,
            expected: 'kitty'
          },
          action: {
            kind: ActionEnum.NAVIGATE,
            fieldId: 'id_number'
          }
        }
      ]
    }
  ]

  const result = applyLogicToFields(_fields, _logics, undefined, {
    id_short_text: 'hello kitty'
  })
  expect(result.fields.map(f => f.id)).toStrictEqual(
    _fields.filter(f => f.id !== 'id_statement').map(f => f.id)
  )
  expect(result.fields).toMatchSnapshot()

  const result2 = applyLogicToFields(_fields, _logics, undefined, {
    id_short_text: 'hello world'
  })
  expect(result2.fields[0].isTouched).toBe(true)

  const result3 = applyLogicToFields(_fields, _logics, undefined, {
    id_short_text: undefined
  })
  expect(result3.fields[0].isTouched).toBe(false)

  const result4 = applyLogicToFields(_fields, _logics, undefined, {
    id_short_text: ''
  })
  expect(result4.fields[0].isTouched).toBe(false)
})

test('apply to fields with thank you page', () => {
  const fields: FormField[] = [
    {
      id: 'id_short_text',
      title: 'Short Text',
      kind: FieldKindEnum.SHORT_TEXT
    },
    {
      id: 'id_statement',
      title: 'Statement',
      kind: FieldKindEnum.STATEMENT
    },
    {
      id: 'id_thank_you',
      title: 'Thank You',
      kind: FieldKindEnum.THANK_YOU
    }
  ]

  const logics: Logic[] = [
    {
      fieldId: 'id_short_text',
      payloads: [
        {
          id: 'p1',
          condition: {
            comparison: ComparisonEnum.CONTAINS,
            expected: 'kitty'
          },
          action: {
            kind: ActionEnum.NAVIGATE,
            fieldId: 'id_thank_you'
          }
        }
      ]
    }
  ]

  const result = applyLogicToFields(fields, logics, undefined, {
    id_short_text: 'hello kitty'
  })

  expect(result.fields.length).toBe(1)
  expect(result.fields[0].id).toBe(fields[0].id)
})
