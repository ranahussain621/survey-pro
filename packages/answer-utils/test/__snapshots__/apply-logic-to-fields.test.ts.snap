// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply to group fields 1`] = `
Array [
  Object {
    "id": "id_short_text",
    "index": 1,
    "isTouched": true,
    "kind": "short_text",
    "title": "Short Text",
  },
  Object {
    "id": "id_group",
    "index": 2,
    "kind": "group",
    "title": "Group",
  },
  Object {
    "id": "id_number",
    "index": 1,
    "kind": "number",
    "parent": Object {
      "id": "id_group",
      "index": 2,
      "kind": "group",
      "title": "Group",
    },
    "title": "Number",
  },
  Object {
    "id": "id_statement2",
    "kind": "statement",
    "parent": Object {
      "id": "id_group",
      "kind": "group",
      "title": "Group",
    },
    "title": "Statement2",
  },
  Object {
    "id": "id_date",
    "index": 3,
    "kind": "date",
    "title": "Date",
  },
]
`;
